src/infra/ChatAnalyticsReporter.ts

```diff
@@ -0,0 +1,74 @@
+import type { EventPayload } from '../models/chat';
+import type ChatAnalyticsConfig from '../models/config/ChatAnalyticsConfig';
+
+interface EventsAPIResponse {
+  id: string;
+}
+
+/**
+ * A class to report chat analytics data. Uses the provided API key, environment,
+ * and partition to determine the endpoint for reporting events.
+ */
+export default class ChatAnalyticsReporter {
+  public readonly apiKey: string;
+  public readonly env: NonNullable<ChatAnalyticsConfig['env']>;
+  public readonly partition: NonNullable<ChatAnalyticsConfig['partition']>;
+  public readonly endpoint: string;
+
+  private endpoints: Record<
+    NonNullable<ChatAnalyticsConfig['partition']>,
+    Record<NonNullable<ChatAnalyticsConfig['env']>, string>
+  > = {
+    US: {
+      PROD: 'https://www.us.yextevents.com/accounts/me/events',
+      SANDBOX: 'https://www.sbx.us.yextevents.com/accounts/me/events',
+      QA: 'https://www.qa.us.yextevents.com/accounts/me/events',
+    },
+    EU: {
+      PROD: 'https://www.eu.yextevents.com/accounts/me/events',
+      SANDBOX: 'https://www.sbx.eu.yextevents.com/accounts/me/events',
+      QA: 'https://www.qa.eu.yextevents.com/accounts/me/events',
+    }
+  };
+
+  constructor({ apiKey, env, partition }: ChatAnalyticsConfig) {
+    this.apiKey = apiKey;
+    this.env = env ?? 'PROD';
+    this.partition = partition ?? 'US';
+    this.endpoint = this.endpoints[this.partition][this.env];
+  }
+
+  /**
+   * Report an event to the chat analytics API.
+   * @param event The event to report.
+   * @returns A promise that resolves to the response from the API.
+   * @example
+   * reporter.report({
+   *  action: 'DIRECTIONS',
+   *   chat: {
+   *   botId: 'davish-playground',
+   * },
+   * });
+   * // Response
+   * {
+   *  id: '12345',
+   * }
+  */
+  public async report(event: EventPayload): Promise<EventsAPIResponse> {
+    const res = await fetch(this.endpoint, {
```

Comment:
When using the Fetch API for analytics events, passing keepalive=true seems to be the recommended approach to make sure that the event isn't cancelled by closing the page. https://developer.mozilla.org/en-US/docs/Web/API/fetch#keepalive

But, not all browsers support keepalive yet (mainly Firefox seems to be the straggler).

There's an older API called the Beacon API which is also provided for analytics events; we made sure the new Analytics Events API was compatible with Beacon (you just pass Authorization in the body instead of as a header).  https://developer.mozilla.org/en-US/docs/Web/API/Navigator/sendBeacon

Reply:
Hm. Which do you think we should use?